### PR TITLE
pallet-gear: Host functions weights

### PR DESCRIPTION
**Pull request by ukint-vs**
Host functions benchmarks:
 ```rust
	fn alloc(r: u32, ) -> Weight;
	fn gas(r: u32, ) -> Weight;
	fn gr_gas_available(r: u32, ) -> Weight;
	fn gr_msg_id(r: u32, ) -> Weight;
	fn gr_origin(r: u32, ) -> Weight;
	fn gr_program_id(r: u32, ) -> Weight;
	fn gr_source(r: u32, ) -> Weight;
	fn gr_value(r: u32, ) -> Weight;
	fn gr_value_available(r: u32, ) -> Weight;
	fn gr_size(r: u32, ) -> Weight;
	fn gr_read(r: u32, ) -> Weight;
	fn gr_read_per_kb(n: u32, ) -> Weight;
	fn gr_block_height(r: u32, ) -> Weight;
	fn gr_block_timestamp(r: u32, ) -> Weight;
	fn gr_send_init(r: u32, ) -> Weight;
	fn gr_send_push(r: u32, ) -> Weight;
	fn gr_send_push_per_kb(n: u32, ) -> Weight;
	fn gr_send_commit(r: u32, ) -> Weight;
	fn gr_send_commit_per_kb(n: u32, ) -> Weight;
	fn gr_reply(r: u32, ) -> Weight;
	fn gr_reply_per_kb(n: u32, ) -> Weight;
	fn gr_reply_to(r: u32, ) -> Weight;
	fn gr_debug(r: u32, ) -> Weight;
	fn gr_debug_per_kb(n: u32, ) -> Weight;
	fn gr_exit_code(r: u32, ) -> Weight;
	fn gr_exit(r: u32, ) -> Weight;
	fn gr_leave(r: u32, ) -> Weight;
	fn gr_wait(r: u32, ) -> Weight;
	fn gr_wake(r: u32, ) -> Weight;
	fn gr_create_program_wgas(r: u32, ) -> Weight;
```

- refactored `submit_code` and `submit_program` benchmarks
- added `reinstrument` benchmark
- check `code_len` on submitting
- added `outgoing_limit` in `gear-pallet` config
- added proc-macro for debugging `Schedule`
- added `costs::RuntimeCost`, `gas::Token` and token consumption mechanism which should replace usual `charge_gas()` in future
- added `fn gas` to `Env`
- use ref mut instead of `inner.take()` in `LaterExt` which caused unnecessary memory relocation (https://github.com/gear-tech/gear/pull/834/commits/37caae0e46a4ae2a75978b366614de283ec3a708)

Benchmarks result (MBP 2021 M1 Pro 10-core):
```bash
gear-node benchmark --extra --dev --execution=native -p pallet_gear -e print_schedule --no-median-slopes --no-min-squares
```
```rust
HostFnWeights {
        alloc: "3.8 µs",
        gr_gas_available: "462.5 ns",
        gr_msg_id: "639.3 ns",
        gr_origin: "607.7 ns",
        gr_program_id: "604.4 ns",
        gr_source: "600.9 ns",
        gr_value: "593.0 ns",
        gr_value_available: "591.4 ns",
        gr_size: "392.6 ns",
        gr_read: "663.7 ns",
        gr_read_per_byte: "253 ps",
        gr_block_height: "402.4 ns",
        gr_block_timestamp: "456.4 ns",
        gr_send_init: "745.3 ns",
        gr_send_push: "1.5 µs",
        gr_send_push_per_byte: "383 ps",
        gr_send_commit: "790.4 µs",
        gr_send_commit_per_byte: "4.0 ns",
        gr_reply: "787.8 ns",
        gr_reply_per_byte: "11 ps",
        gr_reply_to: "83.7 ns",
        gr_debug: "588.1 ns",
        gr_exit_code: "72.4 ns",
        gr_exit: "26.1 ms",
        gr_leave: "29.3 µs",
        gr_wait: "7.8 µs",
        gr_wake: "242.1 µs",
        gr_create_program_wgas: "807.9 µs",
        gas: "507.4 ns",
    },
```

@gear-tech/dev

### CODE REVIEW

File: pallets/gear/src/weights.rs [hunk 1]
```diff
@@ -855,734 +877,3513 @@ impl<T: frame_system::Config> WeightInfo for GearWeight<T> {
...
+	// Storage: unknown [0x673a3a6d73673a3a2d0300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a2e0300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a2f0300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a300300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a310300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a320300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a330300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a340300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a350300000000000000000000000000000000000000000000] (r:0 w:1)
+	// Storage: unknown [0x673a3a6d73673a3a360300000000000000000000000000000000000000000000] (r:0 w:1)
```
Comment (NikVolf) [inline pallets/gear/src/weights.rs:2715]: figure out how to clear that spam

File: pallets/gear/src/weights.rs [hunk 2]
```diff
@@ -97,2662 +124,988 @@ pub trait WeightInfo {
...
+			// Standard Error: 559_000
+			.saturating_add((188_604_000 as Weight).saturating_mul(r as Weight))
+			.saturating_add(T::DbWeight::get().reads(14 as Weight))
+			.saturating_add(T::DbWeight::get().writes(11 as Weight))
+	}
+	fn gr_send_push_per_kb(n: u32, ) -> Weight {
+		(1_774_077_000 as Weight)
+			// Standard Error: 181_000
+			.saturating_add((40_208_000 as Weight).saturating_mul(n as Weight))
+			.saturating_add(T::DbWeight::get().reads(13 as Weight))
```
Comment (NikVolf) [inline pallets/gear/src/weights.rs:291]: this is nonsense, send push does not do any db reads, nor writes

File: pallets/gear/src/weights.rs [hunk 3]
```diff
@@ -97,2662 +124,988 @@ pub trait WeightInfo {
...
+	}
+	fn gr_send_push(r: u32, ) -> Weight {
+		(1_594_962_000 as Weight)
+			// Standard Error: 559_000
+			.saturating_add((188_604_000 as Weight).saturating_mul(r as Weight))
+			.saturating_add(T::DbWeight::get().reads(14 as Weight))
+			.saturating_add(T::DbWeight::get().writes(11 as Weight))
+	}
+	fn gr_send_push_per_kb(n: u32, ) -> Weight {
+		(1_774_077_000 as Weight)
```
Comment (NikVolf) [inline pallets/gear/src/weights.rs:288]: 1 ms for a syscall that works purely on memory is also non-sense

File: pallets/gear/src/lib.rs [hunk 4]
```diff
@@ -337,6 +343,85 @@ pub mod pallet {
     where
         T::AccountId: Origin,
     {
+        /// Submit program for benchmarks which does not check nor instrument the code.
+        #[cfg(feature = "runtime-benchmarks")]
+        pub fn submit_program_raw(
```
Comment (gshep) [inline pallets/gear/src/lib.rs:348]: Could some common extracted to the new function so `submit_program` & `submit_program_raw` could call it?

**Review by gshep — APPROVED** [review]
There are several tiny remarks nevertheless the work is awesome! 🚀

File: pallets/gear/proc-macro/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
+
+[lib]
+proc-macro = true
+
+[dependencies]
+proc-macro2 = "1"
+quote = "1"
+syn = "1"
```
[the remark this answers is not in the document]
Comment (ukint-vs) [reply pallets/gear/proc-macro/Cargo.toml:20]: it's okay to specify proc macro deps like this

**Comment by NikVolf** [comment]
@breathx add reivew?

**Review by breathx — APPROVED** [review]
Great one!

**Pull request merged**